### DISCUSSION TITLE
Is it possible to iterate over key/value values like a map or dictionary?

### DISCUSSION THREAD

**Discussion by tylerjl** [elkowar/eww#898] [Q&A] [answered]
First of all, _wow_, eww is a fantastic project and it is really a pleasure to work with.

Using [magic vars](https://elkowar.github.io/eww/magic-vars.html) as an example, if I wanted to generate a widget for potentially more than one device (like a battery), I was hoping that I could do something like the following:

```elisp
(box
  (for key, values in EWW_BATTERY
    (label :text "${key} at ${values.capacity}")))
```

My intent here is to alleviate the need to know the name of the keys beforehand, but maybe there's some witchcraft that I can leverage something like `literal` for that I'm not aware of.

Thanks!

**Comment by w-lfchen** [comment]
As far as I'm concerned, this is currently not possible.
The `for`-widget expects JSON arrays if I'm not mistaken, and what you are trying to use are JSON objects.
I don't know whether implementing such functionality for objects is possible or makes sense, as they are unordered by definition, which could cause issues if you are trying to generate a widget from a source that doesn't return the key/value pairs in the same order every time it updates.
If you want to learn more about JSON, check out the [documentation](https://www.json.org/json-en.html).
Now when it comes to your problem, I'd suggest trying to separate the object's key/value pairs into an array of arrays or an array of objects using the `jq` function call in the expression language.
Then you can iterate over the array entries and whatever you have stored in them will be accessible using yuck's expression language syntax.

**Comment by tylerjl** [comment, marked as answer]
Thanks for the input, @W-lfchen! That's helpful context.

For future users, this is very possible via some of the jq filters that are available. Although there isn't necessarily a way to make `(for ...` accept a key/value dictionary directly, you can just reformat json structures in a variety of ways that you can then loop through. For example, acquire the keys to `EWW_BATTERY`:

```
(for battery in {jq(EWW_BATTERY, 'del(.total_avg) | keys')}
  (power :battery battery))
```

...and then index into it:

```
(defwidget power [battery]
  (label :text "${EWW_BATTERY[battery].capacity}%"))
```

Other jq filters like `to_entries` can help as well.
